File: 0515/fail/15684_사다리조작_백준.js
const readline = require('readline')
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
})

input = []
result = -1
dr = [0, 0, 1]
dc = [-1, 1, 0]
solution = () => {
    const nmh = input[0].split(" ")
    n = parseInt(nmh[0])
    m = parseInt(nmh[1])
    h = parseInt(nmh[2])
    ladder = Array.from(Array(h + 2), () => Array(n + 1).fill(false))
    for (let i = 1; i < m + 1; i++) {
        const ab = input[i].split(" ")
        a = parseInt(ab[0])
        b = parseInt(ab[1])
        ladder[a][b] = true
    }
    q = []
    for (let row = 1; row <= h; row++) {
        for (let col = 1; col <= n - 1; col++) {
            if (ladder[row][col] === true ) continue;
            if (ladder[row][col + 1] === true) continue;
            if (col != 1) {
                if (ladder[row][col - 1] == true) continue;
            }
            q.push([row, col])
        }
    }
    
    for (let j = 1; j <= 3; j++){
        if (result !== -1) break;
        for (let i = 0; i < q.length - j + 1; i++){
            comVisit = Array(q.length).fill(false)
            chk = false
            my_com(q, i, 1, [q[i]], j)
            if (chk){
                result = j
            }
            break
        }
    }
    console.log(q)
    console.log(result)
}
my_com = (list, index, k, target, c) => {
    comVisit[index] = true
    if (k === c) {
        for (const rowcol of target) {
            ladder[rowcol[0]][rowcol[1]] = true
        }
        console.log(target)
        // console.log(ladder)
        for (let col = 1; col <= n; col++){
            visit = Array.from(Array(h + 2), () => Array(n + 1).fill(false))
            check = false
            dfs(1, col, col)
            // console.log(check)
            console.log('---------')
            if (!check) return;
        }
        chk = true
        // console.log('chk', chk)
        return
    }
    for (let i = index; i < q.length; i++) {
        if (comVisit[i]) continue;
        const qr = q[i][0]
        const qc = q[i][1]
        console.log([qr, qc])
        console.log([qr, qc - 1])
        console.log(target)
        console.log(target.includes([qr, qc - 1]))
        if (target.includes([qr, qc + 1]) || target.includes([qr, qc - 1])) continue;
        k ++
        comVisit[i] = true
        target.push(q[i])
        my_com(list, i, k, target, c)
        k --
        comVisit[i] = false
        target.pop()
    }
}

dfs = (row, col, destination) => {
    visit[row][col] = true
    // console.log(row, col, destination)
    if (row === h + 1) {
        if (destination === col) {
            check = true
            return     
        }
        return
    }
    for (let w = 0; w < 3; w++){
        let nr = row + dr[w]
        let nc = col + dc[w]
        // console.log(nr, nc)
        // console.log(visit[nr][nc])
        // console.log(w)
        if (nr > h + 1 || nc < 0 || nc > n) continue;
        if (visit[nr][nc] === true) continue;
        if (w === 0) {
            // console.log(ladder)
            // console.log(ladder[nr][nc])
            // console.log('*******')
            if (ladder[nr][nc] === false) continue;
        } else if(w === 1) {
            // console.log(ladder)
            // console.log(ladder[row][col])
            // console.log('*******')
            if (ladder[row][col] === false) continue;
        }
        // console.log('?')
        dfs(nr, nc, destination)
        break
    }
}

rl.on('line', line => input.push(line))
.on('close', () => {
    solution()
    process.exit()
})

File: 7월/0703/15684_사다리조작_백준.py
import sys
sys.stdin = open('15684.txt')


def my_com(n, k, ladder, r, c):
    global possible
    if possible == True:
        return

    if n == k:
        for col in range(1, M):
            if not bfs(col, ladder):
                break
        else:
            possible = True
            return

        return

    for row in range(r, N + 1):
        for col in range(1, M):
            if row == r and col <= c:
                continue
            if ladder[row][col] == 1:
                continue
            if col > 1:
                if ladder[row][col - 1] == 1:
                    continue
            if col < M:
                if ladder[row][col + 1] == 1:
                    continue
            
            ladder[row][col] = 1
            my_com(n + 1, k, ladder, row, col)
            ladder[row][col] = 0


def bfs(start, ladder):
    row = 1
    col = start
    mode = 0 # 0:위에서 내려옴, 1: 옆으로 타고옴
    while True:
        if mode == 0:
            if ladder[row][col] == 1:
                col += 1
                mode = 1
            elif ladder[row][col - 1] == 1:
                col -= 1
                mode = 1
            else:
                row += 1
        else:
            mode = 0
            row += 1

        if row == N + 1:
            if col == start:
                return True
            else:
                return False


M, H, N = map(int, input().split())

ladder = [[0 for _ in range(M + 1)] for _ in range(N + 1)]

for _ in range(H):
    a, b = map(int, input().split())
    ladder[a][b] = 1

possible = False
for num in range(0, 4):
    my_com(0, num, ladder, 1, 0)
    if possible == True:
        print(num)
        break
else:
    print(-1)